### PR TITLE
Use an iterator for reader::tagged_docs.

### PR DESCRIPTION
**Pull request by Ms2ger**

### CODE REVIEW

File: src/librustc/metadata/decoder.rs [hunk 1]
```diff
@@ -233,10 +220,16 @@ fn get_provided_source(d: rbml::Doc, cdata: Cmd) -> Option<ast::DefId> {
     })
 }
 
-fn each_reexport<F>(d: rbml::Doc, f: F) -> bool where
+fn each_reexport<F>(d: rbml::Doc, mut f: F) -> bool where
```
Comment (alexcrichton) [inline src/librustc/metadata/decoder.rs]: Perhaps this could return a `TaggedDocsIterator` instead of continuing to use a closure?

**Comment by alexcrichton** [comment]
I do love me some iterator goodness, thanks @Ms2ger!

**Comment by Manishearth** [comment]
```
/home/rustbuild/src/rust-buildbot/slave/auto-linux-64-x-android-t/build/src/librustc/metadata/decoder.rs:222:39: 222:61 error: use of undeclared type name `TaggedDocsIterator`
/home/rustbuild/src/rust-buildbot/slave/auto-linux-64-x-android-t/build/src/librustc/metadata/decoder.rs:222 fn reexports<'a>(d: rbml::Doc<'a>) -> TaggedDocsIterator<'a> {
```

**Comment by Manishearth** [comment]
```
/home/rustbuild/src/rust-buildbot/slave/auto-linux-64-x-android-t/build/src/librustc/metadata/decoder.rs:634:9: 634:13 error: mismatched types:
 expected `()`,
    found `bool`
(expected (),
    found bool) [E0308]
/home/rustbuild/src/rust-buildbot/slave/auto-linux-64-x-android-t/build/src/librustc/metadata/decoder.rs:634         true
                                                                                                                     ^~~~
note: in expansion of for loop expansion
/home/rustbuild/src/rust-buildbot/slave/auto-linux-64-x-android-t/build/src/librustc/metadata/decoder.rs:604:5: 635:6 note: expansion site
/home/rustbuild/src/rust-buildbot/slave/auto-linux-64-x-android-t/build/src/librustc/metadata/decoder.rs:634:9: 634:13 help: run `rustc --explain E0308` to see a detailed explanation
error: aborting due to previous error
make: *** [x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/stamp.rustc] Error 101
```

**Comment by steveklabnik** [comment]
this caused https://github.com/rust-lang/rust/pull/25870 to fail, i think

**Pull request merged**